src/python/pants/backend/go/util_rules/third_party_pkg.py

```diff
@@ -327,23 +330,77 @@ async def _check_go_sum_has_not_changed(
             )
 
 
+@rule
+async def process_find_go_packages(
+    request: ModulePackageListRequest,
+    bash: BashBinary,
+    find: FindBinary,
+) -> ModulePackageList:
+    """This connects the (cached) go mod download steps with the package analyzer steps by finding
+    likely package directories within a module."""
+
+    # As background, several other options did not work for this layer:
+    # - `go list` provides the same package paths, but fails on modules which do not use go.sum
+    # - `os.walk` could process the named cache folder directly, but running a Process allows caching
```

Comment:
I would also add the point that rule code is generally forbidden from using `os.walk` in any case because it prevents the Pants rule engine from seeing what files were accessed (via the `Digest`'s used).